### PR TITLE
feat(disruption): add node notready controller

### LINKED ISSUE #1659: Node NotReady Disruption Controller
### Description

**What problem are you trying to solve?**
Sometimes nodes just become `NotReady` for a variety of reasons (bad cloud provider instance, non-responsive kubelet, etc). When a Node has been in a `Ready` state and then transitions into `NotReady`, I think that Karpenter should have another Disruption Controller that monitors for these nodes and terminates them.

Third party controllers like the Spot.io Ocean Product, and the Cluster Autoscaler both handle nodes that become `NotReady` for you automatically. Karpenter should be able to do the same thing.

(_Note we have also raised this with our AWS TAM via a support ticket, and we were recommended to open a feature-request here_)

Related: https://github.com/kubernetes-sigs/karpenter/issues/1573

**How important is this feature to you?**

This is actually a blocker for us migrating off of our current tools - we launch enough nodes and we have enough failures throughout the day that we cannot fully migrate unless we have a completely automated self healing system where these nodes get cycled out once they become `NotReady`.

(_separate but related, is the ongoing discussion at https://github.com/bottlerocket-os/bottlerocket/issues/4075 about EKS nodes becoming unready due to heavy memory pressure_)

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by mariuskimmina**
Fixes #1659 

**Description**
We would like karpenter to be able to terminate nodes if they have been in an unreachable state for too long.
This has happened to us in the past and as far as I can tell spotio for example already handles this case. 
We experienced such a case of the node becoming unreachable when the kubelet on the node died.

This pr introduces a new field to the nodepool `unreachableTimeout` which can be set to e.g. 10 minutes so that Karpenter would actively terminate a node when it has been unreachable for more than 10 minutes.

We called it notready controller as that's the state the nodes are in when they become unreachable but there might be a better alternative. 

**How was this change tested?**

We added a test suite for this case and we also tested it on one of our EKS test clusters where we simulated a node becoming unreachable and had Karpenter mark the nodeclaim for deletion.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mariuskimmina** [comment]
I think this does count as corporate contribution, it's the first time our company does it tho, so bare with me while I am trying to figure the CLA stuff out.

File: kwok/apis/crds/karpenter.kwok.sh_kwoknodeclasses.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.16.3
+    controller-gen.kubebuilder.io/version: v0.16.4
```
Comment (sftim) [inline kwok/apis/crds/karpenter.kwok.sh_kwoknodeclasses.yaml:6]: Can we (you) split this bump into its own commit / explain more about the context?

File: pkg/controllers/nodeclaim/notready/controller.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == corev1.TaintNodeUnreachable {
+			if taint.TimeAdded != nil {
+				durationSinceTaint := time.Since(taint.TimeAdded.Time)
+				if durationSinceTaint > *nodeClaim.Spec.UnreachableTimeout.Duration {
+					// if node is unreachable for too long, delete the nodeclaim
+					if err := c.kubeClient.Delete(ctx, nodeClaim); err != nil {
+						log.FromContext(ctx).V(0).Error(err, "Failed to delete nodeclaim", "node", node.Name)
```
Comment (sftim) [inline pkg/controllers/nodeclaim/notready/controller.go:65]: (nit)
```suggestion
						log.FromContext(ctx).V(0).Error(err, "Failed to delete NodeClaim", "node", node.Name)
```
?

And other places in the file.
At HTTP level, the lower case `nodeclaims` name is plural. The API kind NodeClaim uses UpperCamelCase.

File: kwok/apis/crds/karpenter.kwok.sh_kwoknodeclasses.yaml (hunk 1 above)
Comment (mariuskimmina) [reply kwok/apis/crds/karpenter.kwok.sh_kwoknodeclasses.yaml:6]: This bump was created by running `make verify` and then presumably by `go generate ./...` - as this is my first time working on karpenter I wasn't sure if I should commit this change. Happy to remove the bumps if it is deemed not necessary.

File: pkg/controllers/nodeclaim/notready/controller.go (hunk 2 above)
Comment (mariuskimmina) [reply pkg/controllers/nodeclaim/notready/controller.go]: Thank you, adapted the UpperCamelCase NodeClaim in all logs

File: kwok/apis/crds/karpenter.kwok.sh_kwoknodeclasses.yaml (hunk 1 above)
Comment (sftim) [reply kwok/apis/crds/karpenter.kwok.sh_kwoknodeclasses.yaml:6]: (I would make the autogenerated changes a separate commit, then it is easy to omit if appropriate)

File: pkg/controllers/nodeclaim/notready/controller.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		return reconcile.Result{}, nodeclaimutil.IgnoreDuplicateNodeError(nodeclaimutil.IgnoreNodeNotFoundError(err))
+	}
+
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == corev1.TaintNodeUnreachable {
+			if taint.TimeAdded != nil {
+				durationSinceTaint := time.Since(taint.TimeAdded.Time)
+				if durationSinceTaint > *nodeClaim.Spec.UnreachableTimeout.Duration {
+					// if node is unreachable for too long, delete the nodeclaim
+					if err := c.kubeClient.Delete(ctx, nodeClaim); err != nil {
```
Comment (sftim) [inline pkg/controllers/nodeclaim/notready/controller.go:66]: - Should something happen to the `.status` of the NodeClaim before deletion?
- Should we record an Event regarding the NodeClaim with the Node as related? (I would)

**Comment by njtran** [comment]
@mariuskimmina fyi if you haven't seen or were aware of, the @engedaam opened up an RFC that seems to tackle the same set of issues :) https://github.com/kubernetes-sigs/karpenter/pull/1768

**Comment by mariuskimmina** [comment]
> @mariuskimmina fyi if you haven't seen or were aware of, the @engedaam opened up an RFC that seems to tackle the same set of issues :) #1768

@njtran thanks for the heads up, his approach does seem more well thought out - I am not sure how I should proceed from here

* should this PR remain open?
* will @engedaam also take care of the implementation?
* anything else I can do to help?

**Comment by engedaam** [comment]
Hey @mariuskimmina, I'm currently planning on handling the implementation. This is a problem space we are trying to move quickly on to help solve for users. We can close this PR out. If you have the time I would appropriate any and all feedback you can provide on both the RFC and implantation

**Comment by mariuskimmina** [comment]
Closing in favor of https://github.com/kubernetes-sigs/karpenter/pull/1793
